src/nxt_otel.c

```diff
@@ -0,0 +1,356 @@
+        return;
+    }
+
+    nxt_list_each(cur, r->fields) {
+        // we need this in a continguous and null terminated segment of memory for Rust FFI
+        name_cur = nxt_mp_zalloc(r->mem_pool, cur->name_length + 1);
+        val_cur = nxt_mp_zalloc(r->mem_pool, cur->value_length + 1);
+        if (name_cur != NULL && val_cur != NULL) {
+            nxt_cpystrn(name_cur, cur->name, cur->name_length);
+            nxt_cpystrn(val_cur, cur->value, cur->value_length);
+            nxt_otel_add_event_to_trace(r->otel->trace, name_cur, val_cur);
+        }
+    } nxt_list_loop;
+
+    // Add method and path to the trace as well
+    // 1. method first
+    name_cur = val_cur = NULL;
+    name_cur = nxt_mp_zalloc(r->mem_pool, sizeof(NXT_OTEL_METHOD_TAG) + 1);
+    val_cur = nxt_mp_zalloc(r->mem_pool, r->method->length + 1);
+    if (name_cur != NULL && val_cur != NULL) {
+        sprintf((char *) name_cur, NXT_OTEL_METHOD_TAG);
+        nxt_cpystrn(val_cur, r->method->start, r->method->length);
+
+        nxt_otel_add_event_to_trace(r->otel->trace, name_cur, val_cur);
+    }
+
+    // 2. path second
+    name_cur = val_cur = NULL;
+    name_cur = nxt_mp_zalloc(r->mem_pool, sizeof(NXT_OTEL_PATH_TAG) + 1);
+    val_cur = nxt_mp_zalloc(r->mem_pool, r->path->length + 1);
+    if (name_cur != NULL && val_cur != NULL) {
+        sprintf((char *) name_cur, NXT_OTEL_PATH_TAG);
+        nxt_cpystrn(val_cur, r->path->start, r->path->length);
+
+        nxt_otel_add_event_to_trace(r->otel->trace, name_cur, val_cur);
+    }
+
+    traceval = nxt_mp_zalloc(r->mem_pool, NXT_OTEL_TRACEPARENT_LEN + 1);
+    if (traceval == NULL) {
+        /* let it go blank here.
+         * span still gets populated and sent
+         * but data is not propagated to peer or app.
+         */
+        nxt_log(t, NXT_LOG_ERR,
+                "couldnt allocate traceparent header. span will not propagate");
+        return;
+    }
+
+    // if we didnt inherit a trace id then we need to add the
+    // traceparent header to the request
+    if (r->otel->trace_id == NULL) {
+        nxt_otel_copy_traceparent(traceval, r->otel->trace);
+        f = nxt_list_add(r->fields);
+        if (f == NULL) {
+            goto next;
+        }
+
+        nxt_http_field_name_set(f, "traceparent");
+        f->value = traceval;
+        f->value_length = nxt_strlen(traceval);
+        nxt_otel_add_event_to_trace(r->otel->trace, f->name, traceval);
+    } else {
+        // copy in the pre-existing traceparent for the response
+        snprintf((char *) traceval, NXT_OTEL_TRACEPARENT_LEN + 1, "%s-%s-%s-%s",
+                 (char *) r->otel->version,
+                 (char *) r->otel->trace_id,
+                 (char *) r->otel->parent_id,
+                 (char *) r->otel->trace_flags);
+    }
+
+    f = nxt_list_add(r->resp.fields);
+    if (f == NULL) {
+        nxt_log(t, NXT_LOG_ERR,
+                "couldnt allocate traceparent header in response");
```

Comment:
I specifically did not want this to return failure. This procedure can be somewhat fault tolerant by omitting some functionality. In this case I think a `goto next;` would allow us to at least move forward with the OpenTelemetry span than returning error.

What do you think? Is there a reason we should prefer returning error?

Reply:
If error happens like in `nxt_otel_span_add_headers()`, it's ok to be tolerant, but it doesn't make sense to continue tracing in later functions like `nxt_otel_span_add_body()`, right?

> Is there a reason we should prefer returning error?

APIs in `nxt_otel.c` looks like to be used by http part, so it's better that the APIs are complete and clear, if errors happen, I'd like to know it through the return type. And if strictly treat it as an error, it's upper to the caller. For example:
```
static nxt_int_t
nxt_otel_span_add_headers(nxt_task_t *task, nxt_http_request_t *r)
{
    if (...) {
        return NXT_ERROR;
    }

    return NXT_OK;
}

ret = nxt_otel_span_add_headers(...);
if (ret) { ...}
Or 
(void) nxt_otel_span_add_headers(...);
```
From the current code, developers have to understand it through something like docs or comments.

Reply:
> APIs in `nxt_otel.c` looks like to be used by http part, so it's better that the APIs are complete and clear, if errors happen, I'd like to know it through the return type. And if strictly treat it as an error, it's upper to the caller. For example:

OK, so maybe there is _no_ C API really. Rust exports some functions and we call them.

Would renaming `nxt_otel.c` => `nxt_http_otel.c` help?

Reply:
nxt_http_otel.c looks more intuitive to me.

Reply:
If APIs with prefix nxt_otel are exported by rust stuff, I feel we could consider naming it nxt_http_trace. “trace” is an action in OTEL and we used it here, right?

Reply:
trace is one of the three things telemetry has a concept of. Metrics and logs are the other two. Ideally the otel capabilities would handle all three, so renaming the file to nxt_http_trace.c would not be accurate

Reply:
Do we use metric and log in http part?

Reply:
Actually it depends on what the file provides. If we change the file name, the function names will be changed.
Which name do you prefer?

Reply:
(Rambling rant time...)

Namings hard... (and I'm still trying to get my head around just what exactly this is all for...)

If there was a C library for OTEL that we we're using, I could see us having a `nxt_otel.c` that provided a slightly higher level interface on-top of that for use in Unit.

I would perhaps try and limit its direct knowledge of Unit internals and limit it to just the bits needed to initialise it and sending the actual data.

It perhaps wouldn't know about Unit requests and fields etc, this would be handled at the place where this is actually being used. Perhaps you create other source files named for such things...

If we remove the configuration validation out from `nxt_otel.c` there isn't much left.

If we implement something like I suggested for where we are creating new heap allocated objects just for passing into `nxt_otel_add_event_to_trace()` only to be copied (more heap allocations), then that gets rid of a fair few lines of code (and unnecessary memory allocations).

Then some of the remaining functions could perhaps be uplifted into where they are actually used, or if they are need in multiple TU's well, we can figure something out... (still familiarising myself with the code...)

Maybe the stuff under `src/otel` is that _"C library"_ and the above can still apply...

Reply:
@hongzhidao 

> If error happens like in nxt_otel_span_add_headers(), it's ok to be tolerant, but it doesn't make sense to continue tracing in later functions like nxt_otel_span_add_body(), right?

For this specific error case I disagree. This logic is placing a traceparent header in response, so the tracing information will not be passed out to the caller by headers in the response, but the span for this request has already been generated. The rest of the calls like `nxt_otel_span_add_body` are attaching information to that span. I think it is best to let the HTTP trace processing continue because the span can still be filled out and sent.

Reply:
If errors have happened, why can the later trace be successful?
For example, memory allocation failed at some point, it means it's an internal server error, it's better not to do the same thing again.

Reply:
> APIs in nxt_otel.c looks like to be used by http part, so it's better that the APIs are complete and clear, if errors happen, I'd like to know it through the return type. And if strictly treat it as an error, it's upper to the caller. For example:

As it stands right now this is a very low coupling design that allows for us to manage errors in the OpenTelemetry module without changing any logic in request processing. Your proposal is to move OpenTelemetry error processing into the calling modules, which spreads it out between the HTTP request processing functions as well as the h1proto connection management functions. This is much less organized I think. 

Also, the logic will be the same. The same errors will be logged and bypassed, but it will be much harder to follow control flow, since we will be spreading all the logic out across two separate state machines that do not have always have predictable state.

I think it is best to leave the design as is. We have one single OpenTelemetry state machine which can manage its own state and control flow, and it only needs to be tested when ready by request and connection processing.

Reply:
> If we remove the configuration validation out from `nxt_otel.c` there isn't much left.

Configuration validation is three small functions in this file. Most of the logic of this module is in header processing and data generation.

Reply:
> If errors have happened, why can the later trace be successful?
>  For example, memory allocation failed at some point, it means it's an internal server error, it's better not to do the same thing again.

This function can return multiple different errors. None of the error cases here are fatal. The Span is already allocated and associated with an ID before this function is called. If memory allocation failures happen here the header processing logic will be interrupted and an error will be logged.

Later, which the body information is added, it is possible that more memory may be available (for example if a request has finished in the meantime). Either way, the body function can handle its own errors. 

In this module there are fatal errors. These errors are handled by logging and switching the otel status to an error status, where any future call to the state machine terminates. See `nxt_otel_test_and_call` for more information.

Reply:
I guess Andrew's thought is for now `nxt_otel.c` looks like a otel library which is used as a callee, but the file includes the `nxt_http_request_t` which is a caller. And I have the same confusion.

Reply:
A bit of techno babble but; I think it's about having correct interface boundaries and avoiding layering violations...

ZFS is an example of a "total layering violation"... It does way too many things incorporating the functionality of (using Linux speak) a filesystem and md and lvm and probably other stuff.

Reply:
I dont think there is a layering violation here. The OpenTelemetry module provides its own state machine that handles all opentelemetry related logic so that none of it has to be encoded in the protocol or request decoding state machines. In this way the current design mimics and actually provides a cleaner interface than the one that exists between the protocol module and the request module.

Reply:
> > If we remove the configuration validation out from `nxt_otel.c` there isn't much left.
> 
> Configuration validation is three small functions in this file. Most of the logic of this module is in header processing and data generation.

These functions have **zero** interaction with the OTEL API. They most definitely should go into `nxt_conf_validation.c`.

Even if they did need to query the OTEL API, it's perfectly OK to call out from the conf validation to other parts of unit such as we do for validating network stuff.